docs/source/java/memory.rst

```diff
@@ -0,0 +1,114 @@
+and provides low-level interfaces for working with the contents, similar to ByteBuffer.
+
+Unlike (Direct)ByteBuffer, it has reference counting built in (see the next section).
+
+Reference counting
+==================
+
+Is a technique to help computer programs manage memory. Tracks the reference/pointers to an object, it increase
++1 or decrease -1 the reference counting between the objects.
+
+If an object ValidityBuffer has a reference with object IntVector, then, IntVector should increase the
+reference counting to 1 (0 + 1 = 1), then if at the same time, ValueBuffer has a reference with IntVector,
+then, IntVector should increase the reference counting to 2 (1 + 1 = 2).
+
+.. code-block::
+
+    |__ A = Allocator
+    |____ B = IntVector (reference count = 2 )
+    |____________ ValidityBuffer
+    |____________ ValueBuffer
+    |____ C = VarcharVector (reference count = 2 )
+    |____________ ValidityBuffer
+    |____________ ValueBuffer
+
+Base on best practices at some point you are going to close your allocator objects using ``close()`` method,
+allocators check for reference counting and throw an exception if they are in use.
+
+Reference Manager manages the reference counting for the underlying memory chunk.
+
+Allocators
+==========
+
+One of the interfaces defined by memory-core is BufferAllocator. This interface collect all the definitions for deal
+with byte buffer allocation.
+
+The concrete implementation of the allocator is Root Allocator. Applications should generally
+create one allocator at the start of the program..
+
+Arrow provides a tree-based model for memory allocation. The RootAllocator is created first,
+then all allocators are created as children of that allocator. The RootAllocator is responsible
+for being the master bookkeeper for memory allocations.
+
+As an example of child allocator consider `Flight Client`_ creation.
+
+Memory Modules
+==============
+
+Applications should depend on memory-core and one of the two implementations,
+else an exception will be raised at runtime.
+
+Development Guidelines
+======================
+
+* Use the BufferAllocator interface in APIs instead of RootAllocator.
+* Applications should generally create one allocator at the start of the program.
+* Remember to close() allocators after use (whether they are child allocators or the RootAllocator), either manually or preferably via a try-with-resources statement.
+* Allocators will check for outstanding memory allocations when closed, and throw an exception if there are allocated buffers, this helps detect memory leaks.
+* Allocators have a debug mode, that makes it easier to figure out where a leak originated (Consider to add this parameter to your application: -Darrow.memory.debug.allocator=true)
+* Arrow modules use logback to collect logs configure that properly to see your logs (create logback-test.xml file on resources folder).
```

Comment:
I think this can also go as part of a section about debugging memory leaks/allocations

Reply:
Added Debugging Memory Leaks/Allocation